src/core/socket.c

```diff
@@ -1461,11 +1463,13 @@ static int socket_determine_selinux_label(Socket *s, char **ret) {
                 if (!UNIT_ISSET(s->service))
                         goto no_label;
 
-                c = SERVICE(UNIT_DEREF(s->service))->exec_command[SERVICE_EXEC_START];
+                service = SERVICE(UNIT_DEREF(s->service));
+                c = service->exec_command[SERVICE_EXEC_START];
                 if (!c)
                         goto no_label;
 
-                r = mac_selinux_get_create_label_from_exe(c->path, ret);
+                path = prefix_roota(service->exec_context.root_directory, c->path);
```

Comment:
i wonder if chase_symlinks() wouldn't be the better option here... i.e. follow symlinks here properly instead of letting the kernel do it, so that we respect the root directory as needed.

Reply:
I implemented this in #8591.

Testing was somewhat basic... Not sure if you had some corner case in mind. Let me know if you want me to check something or other, happy to do so.